### PR TITLE
feat: Update dependencies and toolchain versions

### LINKED ISSUE #1854: Upgrade to stablechain
# Goal
The goal of this issue is to update the rust toolchain to the stable channel.
Additionally, nightly could be used instead of a specific dated version--GitHub workflows that use ci-base-image cannot be updated to use nightly until a new version of ci-base-image is published (see Follow-up)

# Changes
- rust-toolchain.toml updated channel to be stable
- Makefile targets updated to use nightly instead of nightly-2023-07-13
- ci-base-image.dockerfile use nightly instead of nightly-2023-07-13
- Add RUST_VERSION to keep track of the rust toolchain
- merge-pr.yml update ci-base-image version to 1.2.0

# How to Test
Make sure all tests pass. [There are no code changes, only toolchain and workflow updates]
Test make format and make lint which now use nightly
# Follow-up
GitHub workflows should be updated to use nightly once a new ci-base-image is published
Cargo.lock may be updated by cargo update
## DEVELOPER NOTE: Check your rustc --version and make sure you are on the latest (1.76) rustup update to get the latest.

### PR DESCRIPTION
**Pull request by mattheworris**
# Goal
The goal of this PR is to update the rust toolchain to the `stable` channel and publish a new version of `ci-base-image`. 
GitHub workflows that use `ci-base-image` cannot be updated to use `nightly-2024-02-04` until a new version of `ci-base-image` is published (see Follow-up)

Closes #1854

# Changes
- `rust-toolchain.toml` updated channel to be `stable`
- `Makefile` targets updated to use `nightly-2024-02-04` instead of `nightly-2023-07-13`
  - `lint` no longer needs `nightly` unstable features for clippy
- `ci-base-image.dockerfile` use `nightly-2024-02-04` instead of `nightly-2023-07-13`
  - Add `RUST_VERSION` to keep track of the rust toolchain
- `merge-pr.yml` update `ci-base-image` version to `1.2.0` and add `RUST_VERSION`

# How to Test
- Make sure all tests pass. [There are no code changes, only toolchain and workflow updates]
- Test `make format` and `make lint` which use `nightly`

# Follow-up Issue #1896 
- GitHub workflows should be updated to use `nightly-2024-02-04` and `RUST_TOOLCHAIN` should be updated to `stable` once a new `ci-base-image` is published
- `Cargo.lock` may be updated by `cargo update`

## DEVELOPER NOTE: Check your `rustc --version` and make sure you are on the latest (1.76) `rustup update` to get the latest.

# Checklist
- [ ] Chain spec updated
- [ ] Custom RPC OR Runtime API added/changed? Updated js/api-augment.
- [ ] Design doc(s) updated
- [ ] Tests added
- [ ] Benchmarks added
- [ ] Weights updated

### CODE REVIEW

**Review by saraswatpuneet — APPROVED** [review]

**Review by enddynayn — APPROVED** [review]
LGTM 👍

File: Makefile [hunk 1]
```diff
@@ -78,13 +78,13 @@ specs-rococo-local:
 
 .PHONY: format
 format:
-	cargo +nightly-2023-07-13 fmt
+	cargo +nightly fmt
```
Comment (wilwade) [inline Makefile:81]: Do we need to make this change over on the other github workflows? (codecov, release, and verify-pr-commit)

Those are all still constrained to this particular version.

File: rust-toolchain.toml [hunk 2]
```diff
@@ -1,5 +1,5 @@
 [toolchain]
-channel = "1.70.0"
+channel = "stable"
```
Comment (wilwade) [inline rust-toolchain.toml:2]: Do we also need to change `RUST_TOOLCHAIN` in `release.yml`? (It is still set to 1.70.0)

**Review by wilwade — COMMENTED** [review]
Two questions, but I think everything else is good.

**Review by wilwade — APPROVED** [review]
lgtm!

- Reviewed code
- Local check

File: rust-toolchain.toml (hunk 2 above)
Comment (mattheworris) [reply rust-toolchain.toml:2]: Also in follow up pr.

**Pull request merged**